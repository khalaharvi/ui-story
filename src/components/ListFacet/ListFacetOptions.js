import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withTheme } from 'emotion-theming';

import ListFacet from './ListFacet';
import ListFacetOptionsList from './ListFacetOptionsList';
import ListFacetOptionsListItem from './ListFacetOptionsListItem';
import ListFacetShowMoreButton from './ListFacetShowMoreButton';

const itemsPropType = PropTypes.arrayOf(
  PropTypes.shape({
    value: PropTypes.any,
    label: PropTypes.node.isRequired,
    items: (...args) => itemsPropType(...args)
  })
);

class ListFacetOptions extends Component {
  static propTypes = {
    cx: PropTypes.func.isRequired,
    items: itemsPropType,
    renderItem: PropTypes.func.isRequired,
    showMore: PropTypes.bool,
    limit: PropTypes.number,
    showMoreLimit: PropTypes.number,
    canRefine: PropTypes.bool,
    className: PropTypes.string
  };

  constructor() {
    super();

    this.state = {
      extended: false
    };
  }

  onShowMoreClick = () => {
    this.setState(state => ({
      extended: !state.extended
    }));
  };

  getLimit = () => {
    const { limit, showMoreLimit } = this.props;
    const { extended } = this.state;
    return extended ? showMoreLimit : limit;
  };

  renderItem = (item) => {
    const items = item.items && (
      <ListFacetOptionsList>
        {item.items
          .slice(0, this.getLimit())
          .map(child => this.renderItem(child, item))}
      </ListFacetOptionsList>
    );

    return (
      <ListFacetOptionsListItem
        className={this.props.cx(
          'option',
          item.isRefined && 'option--active',
          item.noRefinement && 'option--noRefinement',
          items && 'option--parent'
        )}
        key={item.key || item.label}
        isRefined={item.isRefined}
        noRefinement={item.noRefinement}
        parent={!!items}
      >
        {this.props.renderItem(item)}
        {items}
      </ListFacetOptionsListItem>
    );
  };

  renderShowMore() {
    const { showMore, cx } = this.props;
    const { extended } = this.state;
    const disabled = this.props.limit >= this.props.items.length;
    if (!showMore) {
      return null;
    }

    return (
      <ListFacetShowMoreButton
        disabled={disabled}
        className={cx('option--show-more', disabled && 'option--show-more--disabled')}
        onClick={this.onShowMoreClick}
      >
        View {extended ? 'Less' : 'More'}
      </ListFacetShowMoreButton>
    );
  }

  render() {
    const {
      cx,
      items,
      className,
      canRefine
    } = this.props;

    const rootClassName = classnames(cx(''), className);

    if (items.length === 0) {
      return <ListFacet className={rootClassName} canRefine={canRefine}>No results</ListFacet>;
    }

    // Always limit the number of items we show on screen, since the actual
    // number of retrieved items might vary with the `maxValuesPerFacet` config
    // option.
    return (
      <ListFacet className={rootClassName} canRefine={canRefine}>
        <ListFacetOptionsList canRefine={canRefine}>
          {items
            .slice(0, this.getLimit())
            .map(item => this.renderItem(item))}
        </ListFacetOptionsList>
        {this.renderShowMore()}
      </ListFacet>
    );
  }
}

export default withTheme(ListFacetOptions);
