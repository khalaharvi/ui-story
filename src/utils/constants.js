export const PRIMARY_CATEGORY_KEY = 'categories.lvl0';
export const SECONDARY_CATEGORY_KEY = 'categories.lvl1';
export const TERTIARY_CATEGORY_KEY = 'categories.lvl2';
