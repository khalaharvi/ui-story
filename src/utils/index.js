export { default as createClassNames } from './createClassNames';
export { default as getDisplayName } from './getDisplayName';
export { default as capitalizeFirstLetter } from './capitalizeFirstLetter';
export { default as slugify } from './slugify';
