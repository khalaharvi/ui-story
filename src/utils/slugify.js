export default (text) => {
  // Use hash map for special characters
  const specialChars = {"à":'a',"ä":'a',"á":'a',"â":'a',"æ":'a',"å":'a',"ë":'e',"è":'e',"é":'e', "ê":'e',"î":'i',"ï":'i',"ì":'i',"í":'i',"ò":'o',"ó":'o',"ö":'o',"ô":'o',"ø":'o',"ù":'o',"ú":'u',"ü":'u',"û":'u',"ñ":'n',"ç":'c',"ß":'s',"ÿ":'y',"œ":'o',"ŕ":'r',"ś":'s',"ń":'n',"ṕ":'p',"ẃ":'w',"ǵ":'g',"ǹ":'n',"ḿ":'m',"ǘ":'u',"ẍ":'x',"ź":'z',"ḧ":'h',"·":'-',"/":'-',"_":'-',",":'-',":":'-',";":'-'}; // eslint-disable-line

  return text.toString().toLowerCase()
    .replace(/\s+/g, '-') // Replace spaces with -
    .replace(/./g, target => specialChars[target] || target) // Replace special characters using the hash map
    .replace(/&/g, '-and-') // Replace & with 'and'
    .replace(/[^\w-]+/g, '') // Remove all non-word chars
    .replace(/--+/g, '-') // Replace multiple - with single -
    .replace(/^-+/, '') // Trim - from start of text
    .replace(/-+$/, ''); // Trim - from end of text
};
