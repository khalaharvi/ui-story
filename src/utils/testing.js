import { object } from 'prop-types';
import { channel, createBroadcast } from 'emotion-theming';
import { shallow, mount } from 'enzyme';
import theme from '@dering-hall/dh-theme';

const getOptions = (options = {}) => {
  const broadcast = createBroadcast(theme);

  return {
    context: {
      [channel]: broadcast
    },
    childContextTypes: {
      [channel]: object
    },
    ...options
  };
};

export const mountWithTheme = (
  Component,
  options = {}
) => mount(Component, { ...getOptions(options) });

export const shallowWithTheme = (
  Component,
  options = {}
) => shallow(Component, { ...getOptions(options) });

export default {
  mountWithTheme,
  shallowWithTheme
};
