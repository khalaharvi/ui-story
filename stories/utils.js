/* eslint-disable import/prefer-default-export */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { InstantSearch } from 'react-instantsearch/dom';
import { connectStateResults } from 'react-instantsearch/connectors';
import { withMedia, MediaQueryProvider } from 'react-media-query-hoc';
import { withTests as withTestsAddon } from '@storybook/addon-jest';
import { ThemeProvider } from 'emotion-theming';
import dhTheme from '@dering-hall/dh-theme';
import testsResults from '../.jest-test-results.json';

const Wrap = ({ index, children }) => (
  <InstantSearch
    appId={process.env.ALGOLIA_ID}
    apiKey={process.env.ALGOLIA_API_KEY}
    indexName={index}
  >
    {children}
  </InstantSearch>
);

Wrap.propTypes = {
  index: PropTypes.string,
  children: PropTypes.node
};

Wrap.defaultProps = {
  index: process.env.ALGOLIA_PRODUCT_INDEX
};

const WrapThemeProvider = ({ children, theme }) => (
  <ThemeProvider theme={theme}>
    <Fragment>
      {children}
    </Fragment>
  </ThemeProvider>
);

WrapThemeProvider.propTypes = {
  theme: PropTypes.object // eslint-disable-line
};

WrapThemeProvider.defaultProps = {
  theme: dhTheme
};

const SearchStateDebug = connectStateResults(({ searchState }) => (
  <div style={{ margin: '1rem 0', fontFamily: 'monospace' }}>
    <pre
      style={{
        background: '#f6f8fa',
        fontSize: '.65rem',
        padding: '.5rem'
      }}
    >
      <strong>searchState</strong> = {JSON.stringify(searchState, null, '\t')}
    </pre>
  </div>
));

const customQueries = {
  small: 'screen and (max-width: 730px)',
  medium: 'screen and (max-width: 970px)',
  large: 'screen and (max-width: 1240px)',
  xlarge: 'screen and (min-width: 1240px)'
};

const MediaQueriesWrapper = ({ children }) => (
  <MediaQueryProvider queries={customQueries}>
    {children}
  </MediaQueryProvider>
);

const OnSmallOnly = withMedia(({ media, children }) => {
  if (!media.small) {
    return null;
  }

  return (
    <Fragment>
      {children}
    </Fragment>
  );
});

const withTests = withTestsAddon({
  results: testsResults
});

export {
  SearchStateDebug,
  MediaQueriesWrapper,
  OnSmallOnly,
  Wrap,
  WrapThemeProvider,
  withTests
};
/* eslint-enable */
