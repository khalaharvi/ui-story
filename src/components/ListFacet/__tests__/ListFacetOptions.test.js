import React from 'react';
import { mountWithTheme } from 'utils/testing';

import ListFacetOptions from '../ListFacetOptions';

describe('<ListFacetOptions />', () => {
  it('renders correctly', () => {
    const wrapper = mountWithTheme(
      <ListFacetOptions
        items={[
          { value: 'white', count: 10, label: 'white' },
          { value: 'black', count: 20, label: 'black' }
        ]}
        cx={() => null}
        renderItem={() => null}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className', () => {
    const wrapper = mountWithTheme(
      <ListFacetOptions
        items={[
          { value: 'white', count: 10, label: 'white' },
          { value: 'black', count: 20, label: 'black' }
        ]}
        cx={() => null}
        renderItem={() => null}
        className="custom-list-facet-options"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly without refinement', () => {
    const wrapper = mountWithTheme(
      <ListFacetOptions
        items={[
          { value: 'white', count: 10, label: 'white' },
          { value: 'black', count: 20, label: 'black' }
        ]}
        cx={() => null}
        renderItem={() => null}
        canRefine={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
