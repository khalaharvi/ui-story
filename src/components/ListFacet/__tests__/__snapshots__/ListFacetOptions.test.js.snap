// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListFacetOptions /> renders correctly 1`] = `
.emotion-6 {
  position: relative;
}

.emotion-4 {
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.emotion-4 ul {
  padding-left: 15px;
  margin-bottom: 0;
}

.emotion-0 {
  display: block;
  position: relative;
  color: #6d6d6a;
  font-family: Monument Grotesk,sans-serif;
  font-size: 19px;
  line-height: 25px;
}

.emotion-0:hover {
  color: #242424;
}

<WithTheme(ListFacetOptions)
  cx={[Function]}
  items={
    Array [
      Object {
        "count": 10,
        "label": "white",
        "value": "white",
      },
      Object {
        "count": 20,
        "label": "black",
        "value": "black",
      },
    ]
  }
  renderItem={[Function]}
>
  <ListFacetOptions
    cx={[Function]}
    items={
      Array [
        Object {
          "count": 10,
          "label": "white",
          "value": "white",
        },
        Object {
          "count": 20,
          "label": "black",
          "value": "black",
        },
      ]
    }
    renderItem={[Function]}
  >
    <Styled(div)
      className=""
    >
      <div
        className="emotion-6 emotion-7"
      >
        <Styled(ul)>
          <ul
            className="emotion-4 emotion-5"
          >
            <Styled(li)
              className={null}
              key="white"
              parent={false}
            >
              <li
                className="emotion-0 emotion-1"
              />
            </Styled(li)>
            <Styled(li)
              className={null}
              key="black"
              parent={false}
            >
              <li
                className="emotion-0 emotion-1"
              />
            </Styled(li)>
          </ul>
        </Styled(ul)>
      </div>
    </Styled(div)>
  </ListFacetOptions>
</WithTheme(ListFacetOptions)>
`;

exports[`<ListFacetOptions /> renders correctly with className 1`] = `
.emotion-6 {
  position: relative;
}

.emotion-4 {
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.emotion-4 ul {
  padding-left: 15px;
  margin-bottom: 0;
}

.emotion-0 {
  display: block;
  position: relative;
  color: #6d6d6a;
  font-family: Monument Grotesk,sans-serif;
  font-size: 19px;
  line-height: 25px;
}

.emotion-0:hover {
  color: #242424;
}

<WithTheme(ListFacetOptions)
  className="custom-list-facet-options"
  cx={[Function]}
  items={
    Array [
      Object {
        "count": 10,
        "label": "white",
        "value": "white",
      },
      Object {
        "count": 20,
        "label": "black",
        "value": "black",
      },
    ]
  }
  renderItem={[Function]}
>
  <ListFacetOptions
    className="custom-list-facet-options"
    cx={[Function]}
    items={
      Array [
        Object {
          "count": 10,
          "label": "white",
          "value": "white",
        },
        Object {
          "count": 20,
          "label": "black",
          "value": "black",
        },
      ]
    }
    renderItem={[Function]}
  >
    <Styled(div)
      className="custom-list-facet-options"
    >
      <div
        className="custom-list-facet-options emotion-6 emotion-7"
      >
        <Styled(ul)>
          <ul
            className="emotion-4 emotion-5"
          >
            <Styled(li)
              className={null}
              key="white"
              parent={false}
            >
              <li
                className="emotion-0 emotion-1"
              />
            </Styled(li)>
            <Styled(li)
              className={null}
              key="black"
              parent={false}
            >
              <li
                className="emotion-0 emotion-1"
              />
            </Styled(li)>
          </ul>
        </Styled(ul)>
      </div>
    </Styled(div)>
  </ListFacetOptions>
</WithTheme(ListFacetOptions)>
`;

exports[`<ListFacetOptions /> renders correctly without refinement 1`] = `
.emotion-6 {
  position: relative;
}

.emotion-4 {
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.emotion-4 ul {
  padding-left: 15px;
  margin-bottom: 0;
}

.emotion-0 {
  display: block;
  position: relative;
  color: #6d6d6a;
  font-family: Monument Grotesk,sans-serif;
  font-size: 19px;
  line-height: 25px;
}

.emotion-0:hover {
  color: #242424;
}

<WithTheme(ListFacetOptions)
  canRefine={false}
  cx={[Function]}
  items={
    Array [
      Object {
        "count": 10,
        "label": "white",
        "value": "white",
      },
      Object {
        "count": 20,
        "label": "black",
        "value": "black",
      },
    ]
  }
  renderItem={[Function]}
>
  <ListFacetOptions
    canRefine={false}
    cx={[Function]}
    items={
      Array [
        Object {
          "count": 10,
          "label": "white",
          "value": "white",
        },
        Object {
          "count": 20,
          "label": "black",
          "value": "black",
        },
      ]
    }
    renderItem={[Function]}
  >
    <Styled(div)
      canRefine={false}
      className=""
    >
      <div
        className="emotion-6 emotion-7"
      >
        <Styled(ul)
          canRefine={false}
        >
          <ul
            className="emotion-4 emotion-5"
          >
            <Styled(li)
              className={null}
              key="white"
              parent={false}
            >
              <li
                className="emotion-0 emotion-1"
              />
            </Styled(li)>
            <Styled(li)
              className={null}
              key="black"
              parent={false}
            >
              <li
                className="emotion-0 emotion-1"
              />
            </Styled(li)>
          </ul>
        </Styled(ul)>
      </div>
    </Styled(div)>
  </ListFacetOptions>
</WithTheme(ListFacetOptions)>
`;
